### ISSUE TITLE
Add API for unquoted escaped strings

### ISSUE THREAD

**Issue by nickshanks** [mysqljs/sqlstring#39]
My first attempt to use Node's mysql package went like this [paraphrasing code]:

```
const mysql = require('mysql');
const result = mysql.query('SELECT * FROM table LIMIT 3');
console.log(result);
```

This worked fine. My second attempt to use Node's mysql package went somewhat downhill:

```
const mysql = require('mysql');
const result = mysql.query('SELECT * FROM table WHERE description LIKE "%?%" LIMIT 3', [req.query.q]);
console.log(result);
```

I traced the problem back to this package. It can be characterised as one of two things:
1) the question mark substitution is context-unaware. It should not wrap quote marks around strings if there already are quote marks around the question mark (i.e. it is preceded by an uneven number of any quote character). Or,
2) There needs to be a second API exposed via node's mysql package which lets programmers perform an unquoted escape (and yes, I've seen #19). This will allow programmers to put escaped, user-sourced values within larger strings without having to strip off the quote marks.

As an aside, I looked at the code of this package and all three points of return from escapeString() duplicitously wrap the escaped string in quote characters, à la `return "'" + val + "'";` — you ought to refactor to have this happen in a single place, e.g.:

```
function escapeString(val) {
  return "'" + escapeUnquotedString(val) + "'";
}

function escapeUnquotedString(val) {
  var chunkIndex = CHARS_GLOBAL_REGEXP.lastIndex = 0;
  var escapedVal = '';
  var match;

  while ((match = CHARS_GLOBAL_REGEXP.exec(val))) {
    escapedVal += val.slice(chunkIndex, match.index) + CHARS_ESCAPE_MAP[match[0]];
    chunkIndex = CHARS_GLOBAL_REGEXP.lastIndex;
  }

  if (chunkIndex === 0) {
    // Nothing was escaped
    return val;
  }

  if (chunkIndex < val.length) {
    return escapedVal + val.slice(chunkIndex);
  }

  return escapedVal;
}
```

This removes duplication and reduces the chances of bugs being introduced (e.g. a new return point being added but forgetting to wrap in quotes).

**Comment by dougwilson** [comment]
> It seems this library is not aware of whether the string to be escaped is intended for use as a plain string, or LIKE or RLIKE value (seems similar to the escape() vs. escapeId() distinction), and always treats values as plain strings, not escaping characters destined for pattern matching specifiers. That's fair enough as long as the next level up (the mysqljs/mysql repo) knows this and sends pattern matching strings to a different escaper function. At the moment it sends everything to this library.

That's right, because this library is not a full sql constructor like https://hiddentao.com/squel/ and is not designed to be; one already exists and it is https://hiddentao.com/squel/ :)

The only difference is at the sql langauge level and at that level there are only two things: values and identifiers. The LIKE etc. takes a value, and even characters like the % can be altered within your sql statement, so without this library fully building out your entire statement, then it would not have any idea what the wildcard chars are for that particular string.

I think, based on what you're describing,  you would be much happier using the https://hiddentao.com/squel/ module over this module anyway, as it is a higher level abstraction, which seems to me what you're desiring here.

**Comment by dougwilson** [comment]
Another thing to note is that this module is also made to function with modules like squel. For example their toParam() method https://hiddentao.com/squel/#parameters produces the two values to pass into this module format. It includes the % character in the array and not in the sql as well, so a change here would break a lot of integration with mysql module and all the various sql formatting / orm libraries.

I'm not saying there cannot be a change here, but ideally they shouldn't all break all of a sudden when they are very useful libs. Probably should open a dialog with all these module as part of initial seeking to determine how to keep them all working together.

**Comment by nickshanks** [comment]
Parties talking to each other never created new problems :-)
I was suggesting a new API rather than breaking backwards compat.
And thanks for the pointer to the squel (pronounced: squeel?) library. Will investigate.
